### PR TITLE
Add Beat Writer script

### PR DESCRIPTION
**Pull request by roryjamesallen**
A script that lets you program 2 channels of drum patterns and add randomisation

### CODE REVIEW

File: software/contrib/beat_writer.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        state = self.load_state_json()
+        self.rhythm_1 = state.get("rhythm_1", self.create_blank_rhythm(self.length))
+        self.rhythm_2 = state.get("rhythm_2", self.create_blank_rhythm(self.length))
+        
+        self.previous_step_ticks = ticks_ms
+        self.previous_step_length = 0
+    
+        self.step = 0
+                
+        self.din_out = cv1
```
Comment (mjaskula) [inline software/contrib/beat_writer.py:24]: [optional] For consistency with the fields below:

```suggestion
        self.din_output = cv1
```

File: software/contrib/beat_writer.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        self.save_state()
+            
+    def check_if_reset_rhythm_2(self):
+        current_ticks = ticks_ms()
+        b2_difference = ticks_diff(current_ticks, b2.last_pressed())
+        if b2_difference > 300:
+            self.reset_rhythm(self.rhythm_2)
+        self.save_state()
+            
+    def reset_rhythm(self, rhythm):
```
Comment (mjaskula) [inline software/contrib/beat_writer.py:97]: [required] This function effectively duplicates the functionality of `create_blank_rhythm`.  I think that we only need one of them. If you got rid of this function, then the code in `check_if_reset_rhythm_2` could use:

```python
if b2_difference > 300:
    self.rhythm_2 = create_blank_rhythm(self.length)
```

File: software/contrib/beat_writer.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            self.rhythm_2_output.value(1)
+        else:
+            return
+        
+        half_beat_length = int(self.previous_step_length / 2)
+        current_time_difference_from_previous_beat = ticks_diff(ticks_ms(), self.previous_step_ticks)
+        if current_time_difference_from_previous_beat > half_beat_length:	#If the length of time since the last beat is more than 1/2 one beat's length
+            rhythm[(self.step + 1) % self.length] = 1 - rhythm[(self.step + 1) % self.length]	#Invert the current step
+        else:
+            rhythm[(self.step - 1) % self.length] = 1 - rhythm[(self.step + 1) % self.length]	#Invert the current step
```
Comment (mjaskula) [inline software/contrib/beat_writer.py:138]:
```suggestion
            rhythm[(self.step - 1) % self.length] = 1 - rhythm[(self.step + 1) % self.length]	#Invert the previous step
```

File: software/contrib/beat_writer.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        
+        self.save_state()
+        
+    def splash_screen(self):
+        oled.centre_text('BEAT WRITER\n\nplug in clock')
```
Comment (mjaskula) [inline software/contrib/beat_writer.py:143]: [optional] Maybe specify where to plug in a clock?

**Review by mjaskula — CHANGES_REQUESTED** [review]
Overall this looks pretty good to me, code wise, though i did't try it out yet in hardware. I added some optional and required code improvements. Let me know what you think.

File: software/contrib/beat_writer.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        return rhythm
+    
+    def update_randomisation(self):
+        self.randomisation_1 = k1.read_position()
+        self.randomisation_2 = k2.read_position()
+    
+    def update_outputs(self):
+        self.update_randomisation()
+        output_1_value = randint(0, 1) if randint(2, 98) < self.randomisation_1 else self.rhythm_1[self.step]
+        output_2_value = randint(0, 1) if randint(2, 98) < self.randomisation_2 else self.rhythm_2[self.step]
```
[the remark this answers is not in the document]
Comment (mjaskula) [reply software/contrib/beat_writer.py:59]: That's fine, I just wanted to make sure that it was intentional. In this case I think that the documentation should be updated to describe this more accurately. As it is now, it reads to me as if it's the probability of a beat.

**Comment by roryjamesallen** [comment]
Outdated

**Pull request closed without merging**